Remove hard syck dependency, make it optional instead

oliverguenther opened the pull request:
This PR tries to remove the hard Syck dependency we currently have. As Syck has been removed a long time, we should no longer force its gem onto new installations.
### Syck was used in
- `db/migrate/20130612120042_migrate_serialized_yaml_from_syck_to_psych.rb`
- `db/migrate/utilities/legacy_journal_migrator.rb`

This PR lets these migrations test whether actual data is being transformed.
When any of these migrations find serialized YAML, we're going to assume that
it MIGHT be coming from Syck.

Thus we add a `LegacyYamler` helper that will dynamically require Syck
and fail with an elaborate message if it hasn't been installed using the
optional bundler group `bundle install --with syck`.

On new installations, both these points are skipped as no data is
retrieved from the queries.

This commit should effectively provide Rails 2.2. compatibility

To install with syck, users can do `bundle install --with syck`.
We may think about removing the legacy migrations at some point, though.

oliverguenther committed:
Remove hard syck dependency, make it optional instead

This includes letting the YAML Syck to Psych migrations test whether
actual data is being transformed.
Some migrations still contain a legacy journal migration helper that MAY depend on
Syck.

When these migrations find serialized YAML, we're going to assume that
it MIGHT be coming from Syck.

Thus we add a `LegacyYamler` method that will dynamically require Syck
and fail with an elaborate message if it hasn't been installed using the
optional bundler group `bundle install --with syck`.

On new installations, both these points are skipped as no data is
retrieved from the queries.

This commit should effectively provide Rails 2.2. compatibility

TeatroIO commented:
I've prepared a stage to preview changes. <a href='http://oliverguenther-fix-syck-migration.opf-openproject-2af22bf12c7dfdad4ddf.ttrcloud.com' target='_blank'>Open stage</a> or <a href='http://oliverguenther-fix-syck-migration.opf-openproject-2af22bf12c7dfdad4ddf.ttrcloud.com/teatro.log' target='_blank'>view logs</a>.

NobodysNightmare commented on .travis.yml:
```diff
@@ -58,6 +58,7 @@ env:
     - "TEST_SUITE=spec:legacy DB=postgres"
 
 before_install:
+  - "gem install bundler"
```
what made this change neccessary?

oliverguenther replied:
The bundler for 2.1 on travis is too old (1.7.X) to parse the `optional: true` group definition in the Gemfile. It requires the latest bundler version 1.10. It doesn't seem to add an awful lot of overhead to the travis build though (1-2 seconds).

NobodysNightmare replied:
probably you should still include comment like that there...

So that future oliver knows:
- why this line was neccessary in the dark past
- and can decide if it is still neccessary in the bright future

NobodysNightmare commented on lib/tasks/migrate_yaml.rake:
```diff
@@ -0,0 +1,87 @@
+    rescue LoadError => e
+      abort = -> (str) { abort("\e[31m#{str}\e[0m") }
+      abort.call <<-WARN
+      You appear to have legacy journals in your previous installation,
```
is this message correct in the context of this task?

NobodysNightmare commented:
:+1: for the idea. Still I did not test it yet, just had a look over the code.

myabc commented:
@oliverguenther this is great new for Ruby 2.2 support - the biggest blockers were problems with syck and being stuck on Rails 3.2. Although I didn't have time to continue working on #2383, once this PR is merged, it should be relatively straight-forward to support Ruby 2.2.

oliverguenther commented:
I've reworked my idea and let both migrations as they were, but only optionally require syck when old YAML data is found in the database. On fresh installations, syck is thus superfluous - and there is no documentation / notice required towards the rake task.

oliverguenther commented:
There are two failing tests that seem to be related to localization. I can reproduce them on `release/4.2` as well, can someone confirm this?

NobodysNightmare commented on db/migrate/migration_utils/legacy_yamler.rb:
```diff
@@ -28,40 +28,37 @@
-    end
+    private
+
+    def load_sych
```
Syck, not Sych ^^

NobodysNightmare commented on db/migrate/migration_utils/legacy_yamler.rb:
```diff
@@ -28,40 +28,37 @@
+    # Tries to load syck and fails with an error
+    # if it was not installed.
+    # To continue with the affected migrations, install syck with `bundle install --with syck`
+    def load_with_sych(yaml)
```
typo also here

NobodysNightmare commented:
Neat.
I like that you could keep the existing migrations and just made them require Syck on demand.

oliverguenther commented:
Thanks @NobodysNightmare for your feedback, I've added a TODO note to the temporary bundler/travis workaround and fixed the typos (boy, I sure got confused after working with Psych vs Syck for a few hours)

oliverguenther commented:
@NobodysNightmare I've rebased this after merging a fix for the broken tests. Travis should agree now.

NobodysNightmare commented:
~~Interesting~~: I needed to clear `tmp/cache`, otherwise I would receive errors in `setting.rb` related to the rails cache.

I don't think that this will be a problem, ~~just found it noteworthy as it was unexpected to me.~~

edit: Of course! I changed from `dev` whcih is on Rails 4 and your PR is on `release/4.2`, which is Rails 3...

NobodysNightmare commented:
Looking good, just migrated a fresh DB from the very beginning to the end without syck on my system :+1:
